### ISSUE TITLE
`devDependencies` linked alongside workspace packages

### ISSUE THREAD

**Issue by mrmckeb** [pnpm/pnpm#5631]
We've hit this internally at Vercel in our frontend monorepo. When using workspace packages, the `devDependencies` of those packages are linked and cause duplicate versions in the target app.

Note that we can solve this by using the same dependency versions throughout the repo as a workaround, but that isn't a scalable solution long-term.

### pnpm version:

7.14.0 and 7.16.0

### Code to reproduce the issue:

Next.js app:
```json
{
  "name": "next-app",  
  "dependencies": {
    "react": "18.0.0",
    "shared": "workspace:*"
  }
}
```

Workspace package:
```json
{
  "name": "shared-workspace",
  "devDependencies": {
    "react": "17.0.0"
  }, 
  "dependencies": {
    "abc": "1.2.3",
  },
  "peerDependencies": {
    "react": "17||18"
  }
}
```

My colleague @codybrouwers created a repro based on the above here:
https://github.com/codybrouwers/pnpm-depedencies

### Expected behavior:

The workspace is linked to the app, but the React dev-dependency is not.

```
next-app/
  node_modules/
    react@18.0.0/
    shared-workspace/
      node_modules/
        abc@1.2.3/
```

### Actual behavior:

The dev-dependency is linked along with the workspace, meaning that two React versions are included in the app, which results in an error.

```
next-app/
  node_modules/
    react@18.0.0/
    shared-workspace/
      node_modules/
        abc@1.2.3/
        react@17.0.0/
```

### Additional information:

Node 16.16
macOS Ventura 13.0

**Comment by NullVoxPopuli** [comment]
I think dependenciesMeta*injected solves this -- is that an option for ya?

I think the behavior should be default. 
But updates to the workspace package don't update the injection

**Comment by mrmckeb** [comment]
Yes, that could be a good workaround for us - obviously at the cost of needing to define that for each shared package, in each app's `package.json`.

But agreed, this feels like a bug and that this should be the default behaviour.

**Comment by NullVoxPopuli** [comment]
Yup, i don't remember if @zkochan had thoughts on a solve for that.

Please excuse the brevity, i'm on my pocket computer atm.

Iirc, the linking that dependenciesMeta*injected does is per file from the store, and that's why hmr breaks. And i think 'by file' linking happens so you can exclude the node_modules from the package directory.

I'm no expert, but my first attempt, if i were implementing this, would be to try to symlink directories as much as possible by analyzing the package.json#exports entries.
Like, if i have everything in a 'dist' folder, that folder shourd be synced always.

I don't see a path forward without 'dist' folder usage. But what do i know.

A change in default behaviors might have to be a pnpm 8 thing, depending on backwards compatibility needs.

**Comment by weyert** [comment]
Any news on this?

**Comment by zkochan** [comment]
Yes, the only solution for this is using injected dependencies.

For [Bit](https://bit.dev/) I have recently implemented a feature called "root components", it is an opt-in feature for now. Bit also uses "injected dependencies" to solve these issues. However, with Bit the dev experience is better because bit cli also handles compilation, so it can automatically keep all the "injected dependencies" up-to-date.

I see there is an attempt to improve this in pnpm via this PR: https://github.com/pnpm/pnpm/pull/6135
